key_test.go

```diff
@@ -58,32 +58,38 @@ func (c *mockCtx) FindObjectsInit(sh pkcs11.SessionHandle, temp []*pkcs11.Attrib
 }
 
 func (c *mockCtx) FindObjects(sh pkcs11.SessionHandle, max int) ([]pkcs11.ObjectHandle, bool, error) {
-	var class uint
-	for _, a := range c.currentSearch {
-		// Hack: we know the relevant classes are all expressible in single bytes.
-		if a.Type == pkcs11.CKA_CLASS {
-			class = uint(a.Value[0])
-		}
-		// We search public keys by CKA_CLASS and CKA_VALUE.
-		// Note that this mock is sensitive to the particular order
-		// in which the attributes are specified in the call to FindObjects.
-		if a.Type == pkcs11.CKA_VALUE && class == pkcs11.CKO_PUBLIC_KEY {
-			switch string(a.Value) {
-			case rsaMarshalled:
-				return []pkcs11.ObjectHandle{rsaPublicKeyHandle}, false, nil
-			case ecMarshalled:
-				return []pkcs11.ObjectHandle{ecPublicKeyHandle}, false, nil
-			}
-		}
-		// We search private keys using CKA_CLASS and CKA_ID
-		if a.Type == pkcs11.CKA_ID && class == pkcs11.CKO_PRIVATE_KEY {
-			switch string(a.Value) {
-			case string([]byte{rsaKeyID}):
-				return []pkcs11.ObjectHandle{rsaPrivateKeyHandle}, false, nil
-			case string([]byte{ecKeyID}):
-				return []pkcs11.ObjectHandle{ecPrivateKeyHandle}, false, nil
-			}
-		}
+	if reflect.DeepEqual(c.currentSearch, []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PUBLIC_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, pkcs11.CKK_RSA),
+		pkcs11.NewAttribute(pkcs11.CKA_MODULUS, rsaKey.N.Bytes()),
+		pkcs11.NewAttribute(pkcs11.CKA_PUBLIC_EXPONENT, big.NewInt(int64(rsaKey.E)).Bytes()),
+	}) {
+		return []pkcs11.ObjectHandle{rsaPublicKeyHandle}, false, nil
+	}
+	if reflect.DeepEqual(c.currentSearch, []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PRIVATE_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_ID, []byte{rsaKeyID}),
+	}) {
+		return []pkcs11.ObjectHandle{rsaPrivateKeyHandle}, false, nil
+	}
+
+	if reflect.DeepEqual(c.currentSearch, []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PUBLIC_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, pkcs11.CKK_EC),
+		pkcs11.NewAttribute(pkcs11.CKA_EC_PARAMS, []uint8{0x6, 0x8, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x3, 0x1, 0x7}),
+		pkcs11.NewAttribute(pkcs11.CKA_EC_POINT, []uint8{0x4, 0x41, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1}),
+	}) {
+		return []pkcs11.ObjectHandle{ecPublicKeyHandle}, false, nil
+	}
+	if reflect.DeepEqual(c.currentSearch, []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PRIVATE_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_ID, []byte{ecKeyID}),
+	}) {
+		return []pkcs11.ObjectHandle{ecPrivateKeyHandle}, false, nil
+	}
+	fmt.Println("unrecognized search:")
+	for _, v := range c.currentSearch {
+		fmt.Printf("  Type: %x, Value: %x\n", v.Type, v.Value)
```

Comment:
I left it here on purpose. When testing and debugging this, I found that the most common test failures were fallthroughs here because I had changed the lookup signature slightly. This print makes it easy to spot what the actual lookup requested.